File: Python/July_problems/problem2181_4th_july_2024.py
from typing import Optional


class ListNode:
    def __init__(self, val=0, next=None):
        self.val = val
        self.next = next


def mergeNodes(self, head: Optional[ListNode]) -> Optional[ListNode]:
    current = head
    answer = ListNode()
    tail = answer
    sum = 0
    while current != None:
        if current.val != 0:
            sum += current.val
        else:
            tail.next = ListNode(sum)
            tail = tail.next
            sum = 0
        current = current.next
    return answer.next


def problem2181():
    pass


if __name__ == '__main__':
    problem2181()

File: CPP/Year-2024/July_Problems/problem2181_4th_july_2024.cpp
#include <iostream>
#include <vector>

using namespace std;

struct ListNode
{
    int val;
    ListNode *next;
    ListNode() : val(0), next(nullptr) {}
    ListNode(int x) : val(x), next(nullptr) {}
    ListNode(int x, ListNode *next) : val(x), next(next) {}
};

ListNode *mergeNodes(ListNode *head)
{
    ListNode *current = head->next;
    ListNode *temp = new ListNode(0);
    ListNode *tail = temp;
    int sum = 0;

    while (current != nullptr)
    {
        if (current->val != 0)
        {
            sum += current->val;
        }
        else
        {
            tail->next = new ListNode(sum);
            tail = tail->next;
            sum = 0;
        }
        current = current->next;
    }

    return temp->next;
}

ListNode *list_to_linked_list(const vector<int> &list)
{
    if (list.empty())
        return nullptr;

    ListNode *head = new ListNode(list[0]);
    ListNode *current = head;

    for (size_t i = 1; i < list.size(); ++i)
    {
        current->next = new ListNode(list[i]);
        current = current->next;
    }

    return head;
}

void print_list(ListNode *head)
{
    ListNode *current = head;
    while (current != nullptr)
    {
        cout << current->val << " ";
        current = current->next;
    }
    cout << endl;
}

int main()
{
    vector<int> list = {0, 3, 1, 0, 4, 5, 2, 0};
    ListNode *head = list_to_linked_list(list);
    ListNode *merged = mergeNodes(head);
    print_list(merged);
    return 0;
}
